BenchmarkVCFs/SimpleBenchmark.wdl

```diff
@@ -0,0 +1,592 @@
+
+            df = df.rename(columns=rename_columns)
+            df['Type'] = Type.upper()
+            df['Stratifier'] = "~{strat_label}"
+            df['Score_Field'] = "~{score_field}"
+            df['Call_Name'] = "~{call_output_sample_name}"
+            df['Base_Name'] = "~{base_output_sample_name}"
+
+            roc_summary = pd.concat([roc_summary, df])
+
+        roc_summary.to_csv('ROC_summary.tsv', sep='\t', index=False)
+
+        CODE
+
+    >>>
+
+    runtime {
+        docker: "us.gcr.io/broad-dsde-methods/rtg:v1.0"
+        preemptible: select_first([preemptible, 0])
+        disks: "local-disk " + disk_size + " HDD"
+        cpu: cpu
+        memory: memory + " GB"
+    }
+
+    output {
+        File ROC_summary = "ROC_summary.tsv"
+
+        File tp_base_vcf = "output_dir/tp-baseline.vcf.gz"
+        File tp_base_index = "output_dir/tp-baseline.vcf.gz.tbi"
+        File tp_call_vcf = "output_dir/tp.vcf.gz"
+        File tp_call_index = "output_dir/tp.vcf.gz.tbi"
+        File fp_vcf = "output_dir/fp.vcf.gz"
+        File fp_index = "output_dir/fp.vcf.gz.tbi"
+        File fn_vcf = "output_dir/fn.vcf.gz"
+        File fn_index = "output_dir/fn.vcf.gz.tbi"
+    }
+}
+
+task BCFToolsStats {
+    input {
+        # Post-vcfeval files
+        File tp_base_vcf
+        File tp_base_index
+        File tp_call_vcf
+        File tp_call_index
+        File fp_vcf
+        File fp_index
+        File fn_vcf
+        File fn_index
+
+        String strat_label
+        String bcf_selector
+        String bcf_label
+
+        String call_output_sample_name
+        String base_output_sample_name
+
+        File? evaluation_bed
+
+        Int disk_size = ceil(size(tp_base_vcf, "GB") + size(tp_call_vcf, "GB") + size(fp_vcf, "GB") + size(fn_vcf, "GB")) + 20
+        Int cpu = 16
+        Int memory = 32
+    }
+
+    # Handle parsing bcf_selector with surrounding quotes, with empty case handled separately
+    String selection = if bcf_selector!= "" then "-i " + '"' + bcf_selector + '"' else ""
+
+    command <<<
+        set -xeuo pipefail
+
+        # Subset to evaluation bed if provided, to ensure not too many FNs picked up outside of it
+        bcftools stats ~{selection} -s- ~{"-R" + evaluation_bed} --threads ~{cpu} ~{tp_base_vcf} > tp_base_stats.tsv
+        bcftools stats ~{selection} -s- ~{"-R" + evaluation_bed} --threads ~{cpu} ~{tp_call_vcf} > tp_call_stats.tsv
+        bcftools stats ~{selection} -s- ~{"-R" + evaluation_bed} --threads ~{cpu} ~{fp_vcf} > fp_stats.tsv
+        bcftools stats ~{selection} -s- ~{"-R" + evaluation_bed} --threads ~{cpu} ~{fn_vcf} > fn_stats.tsv
+
+        python3 << CODE
+        import io
+        import pandas as pd
+
+        def make_dfs(file_path):
+            with open(file_path, 'r') as file:
+                bcf_file = file.readlines()
+
+                # sample-specific stats
+                sn_df = pd.read_csv(io.StringIO('\n'.join([x for x in bcf_file if 'SN\t' in x])), sep='\t')
```

Comment:
Oh I see, you need to access the data again further down. In that case I would argue that it's fine to store the original data and create copies with the transformation in the StringIO, if the files are not large. But then you can take everything out of the `with open…` indentation except for `bcf_file = …`.

Optionally, if you don't want to create multiple copies, you can do what I suggested in the comment above and write `file.seek(0)` every time you need to reset the file iterator to the beginning of the file

Reply:
Yeah I went with "read once, process multiple times" even though it might be a little less efficient than carefully streaming the file and watching for the first column values to change because I also didn't want to depend on the order in the bcftools files themselves.